Automattic/wp-calypso
# Use the site search endpoint for plugin search

### Changes proposed in this Pull Request

* Adds a new hook to use the `https://public-api.wordpress.com/rest/v1/sites/108986944/search` endpoint
* Adds the same parameters for author and category search as WPORG
* Moves some common code to utils and constants files

### Testing instructions



* Visit `/plugins`
* Search for something generic (eg. `wordpress`)
* Verify that both WPCOM and WPORG results are loaded
* Verify that WPORG results have all the necessary information on the search result card (title, excerpt, rating, last update, number of installs)
* Scroll to the bottom, and verify that the search result list loads the next page of results 

**Videos**
Plugin search: https://d.pr/v/bvi2uz
Author search: https://d.pr/v/9ntpmF

### Known issues, next steps

#### Category search

The index contains only WPORG categories, and we can search for those. However, the categories used on the WPORG API (Popular, Featured), are stored in the `plugin_section` taxonomy and are not in the index response: 
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L195
https://github.com/WordPress/wordpress.org/blob/38e38a5b0bbca67affaf0ec70617db14f3f88f84/wordpress.org/public_html/wp-content/plugins/plugin-directory/class-plugin-directory.php#L760

We should assess if we need to add this taxonomy to the index as well.

#### Localization

The current implementation only supports results in english. The index stores all the translations in the MAG16, so we can create the logic of picking the correct keys in a follow-up PR.

#### Icons

The index contains icon file names in a serialized PHP array. These are deserialized and appended to the WPORG image host (`https://ps.w.org`). For eg. Woocommerce plugins, the PHP array is empty, so icon URLs can't be produced just from the index response. We'll need to find an alternative, more stable way of getting the icon URLs.

**To reproduce:** try searching for `WooTcsShipmentMaker`

| Search results (index) | Plugin details page (WPORG) |
| ------ | ----- |
| ![image](https://user-images.githubusercontent.com/11555574/165253817-dfc5ab8b-1394-4789-a717-4d4692a6ffec.png) | ![image](https://user-images.githubusercontent.com/11555574/165253859-668568c0-fb22-43cb-a670-d2da8c6f9630.png) |

#### Author profile link

The index response doesn't contain the `author_profile` URL, which is a WPORG profile URL. I'm not sure we need it though, as the author name currently redirects to the author search page, and not the `author_profile` URL. It's also only used on the plugin details page, not the search results page.

| Search results | Plugin details page |
| ------ | ----- |
| ![image](https://user-images.githubusercontent.com/11555574/165255359-e727067b-ba8b-484a-844e-1e640cc8110e.png) | ![image](https://user-images.githubusercontent.com/11555574/165255640-169b2754-5198-48b3-a65a-f0620c3d832a.png) |

#### Plugin details page

We can query for an exact match on the product slug from the index, which would let us get the same data to display on the plugin details page. However, as the plugin details page uses more fields from the WPORG response, we should investigate in a follow-up issue whether we need all the required fields to display the plugin page.



Related to #62729
